mne/viz/utils.py

```diff
@@ -386,3 +385,110 @@ def figure_nobar(*args, **kwargs):
+
+    """
+
+    def __init__(self, imdata, **kwargs):
+        """Display the image for clicking."""
+        from matplotlib.pyplot import figure, show
+        self.coords = []
+        self.imdata = imdata
+        self.fig = figure()
+        self.ax = self.fig.add_subplot(111)
+        self.ymax = self.imdata.shape[0]
+        self.xmax = self.imdata.shape[1]
+        self.im = self.ax.imshow(imdata, aspect='auto',
+                                 extent=(0, self.xmax, 0, self.ymax),
+                                 picker=True, **kwargs)
+        self.ax.axis('off')
+        self.fig.canvas.mpl_connect('pick_event', self.onclick)
+        show()
+
+    def onclick(self, event):
+        """Mouse click handler."""
+        mouseevent = event.mouseevent
+        self.coords.append((mouseevent.xdata, mouseevent.ydata))
+
+    def plot_clicks(self, **kwargs):
+        """Plot the x/y positions stored in self.coords."""
+        from matplotlib.pyplot import subplots, show
+        f, ax = subplots()
+        ax.imshow(self.imdata, extent=(0, self.xmax, 0, self.ymax), **kwargs)
+        xlim, ylim = [ax.get_xlim(), ax.get_ylim()]
+        xcoords, ycoords = zip(*self.coords)
+        ax.scatter(xcoords, ycoords, c='r')
+        ann_text = np.arange(len(self.coords)).astype(str)
+        for txt, coord in zip(ann_text, self.coords):
+            ax.annotate(txt, coord, fontsize=20, color='r')
+        ax.set_xlim(xlim)
+        ax.set_ylim(ylim)
+        show()
+
+    def to_layout(self, **kwargs):
+        """Turn coordinates into an MNE Layout object.
+
+        Normalizes by the image you used to generate clicks
+        """
+        from mne.channels.layout import generate_2d_layout
+        coords = np.array(self.coords)
+        lt = generate_2d_layout(coords, bg_image=self.imdata, **kwargs)
+        return lt
+
+
+def _fake_click(fig, ax, point, xform='ax'):
+    """Helper to fake a click at a relative point within axes."""
+    if xform == 'ax':
+        x, y = ax.transAxes.transform_point(point)
+    elif xform == 'data':
+        x, y = ax.transData.transform_point(point)
+    else:
+        raise ValueError('unknown transform')
+    try:
+        fig.canvas.button_press_event(x, y, 1, False, None)
+    except:  # for old MPL
+        fig.canvas.button_press_event(x, y, 1, False)
+
+
+def add_background_image(fig, im):
+    """Add a background image to a plot.
+
+    Adds the image specified in `im` to the
+    figure `fig`. This is generally meant to
+    be done with topo plots, though it could work
+    for any plot.
+
+    Note: this sets the aspect ratio of the first
+    axis in "fig" to 'auto'.
+
+    INPUTS
+    --------
+    fig: plt.figure
+        The figure you wish to add a bg image to.
+    im: ndarray
+        A numpy array that works with a call to
+        plt.imshow(im). This will be plotted
+        as the background of the figure.
+    """
+    fig.axes[0].set_aspect('auto')
```

Comment:
Hmmm - what if there was a `set_aspect_ratio` flag, and if if you pass it a
parameter, then it will try:

```python
for ax in fig.axes:
  ax.set_aspect(set_aspect_ratio)
```

or something like this?

On Mon, Apr 27, 2015 at 12:09 PM, Eric Larson <notifications@github.com>
wrote:

> In mne/viz/utils.py
> <https://github.com/mne-tools/mne-python/pull/1762#discussion_r29179732>:
>
> > +    be done with topo plots, though it could work
> > +    for any plot.
> > +
> > +    Note: this sets the aspect ratio of the first
> > +    axis in "fig" to 'auto'.
> > +
> > +    INPUTS
> > +    --------
> > +    fig: plt.figure
> > +        The figure you wish to add a bg image to.
> > +    im: ndarray
> > +        A numpy array that works with a call to
> > +        plt.imshow(im). This will be plotted
> > +        as the background of the figure.
> > +    """
> > +    fig.axes[0].set_aspect('auto')
>
> I'm thinking of this as a general-purpose function to add a background
> image to a figure. A figure in general can have any number of axes, or no
> axes. Thus the specialization it seems should live in the plot_topo
> function. I guess you could have an axes parameter here that accepts an
> axes or list of axes and sets them all to auto aspect ratio if need be...?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/mne-tools/mne-python/pull/1762/files#r29179732>.
>